### DISCUSSION TITLE
Aggregate Tables

### DISCUSSION THREAD

**Discussion by ayuhito** [duckdb/duckdb#11800] [Ideas]
Similar to the [StarRocks](https://docs.starrocks.io/docs/table_design/table_types/aggregate_table/) or [Clickhouse](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/aggregatingmergetree) implementations.

Rather than processing millions of rows, it would be nicer to have some optimised paths to aggregate values in rows partitioned by bounded timestamps. Much more efficient in the case of query speed and storage usage. I understand it doesn't really fit into the DuckDB model, but I'd love to hear opinions if it might work out.

**Comment by Tishj** [comment]
So it's essentially a materialized view, or am I misunderstanding?

The main difference being that materialized views store results of expressions over data stored in a native table, but this doesn't store the data in a native table, only storing the results of the aggregates, right?
And I guess this sort of pretends that the materialized view is the table, for ingestion purposes

Maybe it's closer to having a table with "hidden" columns to store the raw column data, paired with visible physical generated columns  that contain the aggregates over the "hidden" columns

Though by the sounds of it it's using an entirely different storage model for these tables

**Reply by ayuhito** [reply]
Somewhat, yes. A materialized view still stores the original records individually, whilst an aggregate table runs compaction and doesn't store the individual records leading to significantly less data to process and store for related workloads.

They do take a lot from each other  though since both a materialized view and aggregate table would keep itself in sync with the source data. Rather than the "view" aspect though, you're just skipping the step of storing the data of the individual records, only keeping the aggregate.

**Reply by Tishj** [reply]
I'm not sure it's possible to only store the aggregate data for all aggregates, think of an `avg` aggregate for example, we'd need to know the total amount and the amount of values
But yea I can imagine it would be possible to store a smaller amount of data.

**Reply by ayuhito** [reply]
> I'm not sure it's possible to only store the aggregate data for all aggregates

StarRocks only supports this with `SUM`, `MIN`, `MAX` and `REPLACE`. I don't think there are many other aggregate functions you can support with this model.

> think of an avg aggregate for example, we'd need to know the total amount and the amount of values

This can still be done by storing a SUM column and knowing the COUNT for each row. The engine should not support it since there are many variants of queries you could do, but the user can design their own schemas to track it with the already given primitives which would be more than enough. It still can be very useful and versatile.
